File: frontend/demo/component/datepicker/date-picker-date-metadata.ts
import 'Frontend/demo/init'; // hidden-source-line
import '@vaadin/date-picker';
import { eachDayOfInterval, formatISO, parseISO } from 'date-fns';
import { html, LitElement } from 'lit';
import { customElement } from 'lit/decorators.js';
import type { DatePickerDateMetadata, DatePickerDateMetadataProvider } from '@vaadin/date-picker';
import { applyTheme } from 'Frontend/demo/theme';

@customElement('date-picker-date-metadata')
export class Example extends LitElement {
  protected override createRenderRoot() {
    const root = super.createRenderRoot();
    applyTheme(root);
    return root;
  }

  // tag::snippet[]
  // In a real application, these would query a booking service.
  private isFullyBooked = (date: Date) => date.getDate() % 7 === 3;

  private isAlmostFull = (date: Date) => date.getDate() % 5 === 0;

  private getMetadata = (date: Date): DatePickerDateMetadata | undefined => {
    const isoDate = formatISO(date, { representation: 'date' });
    if (this.isFullyBooked(date)) {
      return { date: isoDate, disabled: true };
    }
    if (this.isAlmostFull(date)) {
      return { date: isoDate, part: 'limited' };
    }
    return undefined;
  };

  // Keep a stable reference: assigning a new function clears the cache.
  private dateMetadataProvider: DatePickerDateMetadataProvider = ({ start, end }) =>
    eachDayOfInterval({ start: parseISO(start), end: parseISO(end) })
      .map(this.getMetadata)
      .filter((metadata) => metadata !== undefined);

  protected override render() {
    return html`
      <vaadin-date-picker
        label="Appointment date"
        helper-text="Highlighted dates are almost full"
        .dateMetadataProvider="${this.dateMetadataProvider}"
      ></vaadin-date-picker>
    `;
  }
  // end::snippet[]
}

File: src/main/java/com/vaadin/demo/component/datepicker/DatePickerDateMetadata.java
package com.vaadin.demo.component.datepicker;

import com.vaadin.demo.DemoExporter; // hidden-source-line
import com.vaadin.flow.component.datepicker.DateMetadata;
import com.vaadin.flow.component.datepicker.DateMetadataProvider;
import com.vaadin.flow.component.datepicker.DatePicker;
import com.vaadin.flow.component.datepicker.DatePicker.DatePickerI18n;
import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.router.Route;

import java.time.LocalDate;

@Route("date-picker-date-metadata")
public class DatePickerDateMetadata extends Div {

    public DatePickerDateMetadata() {
        DatePicker datePicker = new DatePicker("Appointment date");
        // tag::snippet[]
        datePicker.setHelperText("Highlighted dates are almost full");

        DateMetadataProvider provider = DateMetadataProvider.perDate(date -> {
            if (isFullyBooked(date)) {
                return new DateMetadata(date, true);
            }
            if (isAlmostFull(date)) {
                return new DateMetadata(date, "limited");
            }
            return null;
        });
        datePicker.setDateMetadataProvider(provider);

        datePicker.setI18n(new DatePickerI18n()
                .setDisabledDateErrorMessage("That date is fully booked"));
        // end::snippet[]

        add(datePicker);
    }

    // tag::snippet[]

    // In a real application, these would query a booking service.
    private static boolean isFullyBooked(LocalDate date) {
        return date.getDayOfMonth() % 7 == 3;
    }

    private static boolean isAlmostFull(LocalDate date) {
        return date.getDayOfMonth() % 5 == 0;
    }
    // end::snippet[]

    public static class Exporter extends DemoExporter<DatePickerDateMetadata> { // hidden-source-line
    } // hidden-source-line
}
